### PR TITLE
ARROW-2019: [JAVA] Control the memory allocated for inner vector in LIST

### PR DESCRIPTION
**Pull request by siddharthteotia**
We have observed cases in our external sort code where the amount of memory actually allocated for a record batch sometimes turns out to be more than necessary and also more than what was reserved by the operator for special purposes. Thus queries fail with OOM.

Usually to control the memory allocated by vector.allocateNew() is to do a setInitialCapacity() and the latter modifies the vector state variables which are then used to allocate memory. However, due to the multiplier of 5 used in List Vector, we end up asking for more memory than necessary. For example, for a value count of 4095, we asked for 128KB of memory for an offset buffer of VarCharVector for a field which was list of varchars. 

We did ((4095 * 5) + 1) * 4 => 80KB . => 128KB (rounded off to power of 2 allocation). 

We had earlier made changes to setInitialCapacity() of ListVector when we were facing problems with deeply nested lists and decided to use the multiplier only for the leaf scalar vector. 

It looks like there is a need for a specialized setInitialCapacity() for ListVector where the caller dictates the repeatedness.

Also, there is another bug in setInitialCapacity() where the allocation of validity buffer doesn't obey the capacity specified in setInitialCapacity(). 

cc @jacques-n , @BryanCutler , @icexelloss , @vkorukanti

### CODE REVIEW

**Comment by jacques-n** [comment]
The change makes sense but I don't understand why the tests change the way they do.

**Comment by siddharthteotia** [comment]
@jacques-n , in addition to the new API of setInitialCapacity(valueCount, multiplier), another bug has been fixed as part of this patch. 

Until now setInitialCapacity() was a part of super class BaseRepeatedValueVector so doing setInitialCapacity() would not control the allocation of validity buffer which is part of sub class ListVector. The call only impacted the offset buffer and data vector since they are members of super class.

So if we do setInitialCapacity(512), it would still allocate validity buffer with default state, i.e for 4096 values. Subsequently, getValueCapacity() does a MIN (offset buffer value capacity - 1, validity buffer capacity) and that's why the result was 1023  -- 

we allocated offset buffer for (512 + 1) * 4 => 2052 bytes => 4096 bytes and thus 1024 value capacity of offset buffer. So the result was 1023.

In other words, getValueCapacity() was previously returning the value capacity of offset buffer. Now since both versions of setInitialCapacity() are implemented in the base class, it correctly reflects the value capacity w.r.t validity buffer.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/BaseRepeatedValueVector.java [hunk 1]
```diff
@@ -143,6 +143,26 @@ public void setInitialCapacity(int numRecords) {
...
+   * such cases, we are likely to face OOM or related problems when
+   * we reserve memory for a record batch with value count x and
+   * do setInitialCapacity(x) such that each vector allocates only
+   * what is necessary and not the default amount but the multiplier
+   * forces the memory requirement to go beyond what was needed.
+   *
+   * @param numRecords value count
+   * @param multiplier repeated value multiplier
+   */
+  public void setInitialCapacity(int numRecords, int multiplier) {
```
Comment (jacques-n) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/BaseRepeatedValueVector.java]: multiplier should be double.

**Comment by jacques-n** [comment]
I gave this some more thought and I think the actual concept is variable width data density. So I think the parameter should probably be called density and we should allow it to be set on varchar/varbinary as well. I think we should also add a new method on each of these types which is getDensity() which returns the relative density of the structure.

For example, density for list vector would be the average list size per entry. Such as:

10 => on average, each position has a list of 10 values.
0.1 => out of ten lists, one has a single element and all the other lists are null.

The same could be applied to varchar/varbinary but the density would be the number of data bytes on average per element. If we have a large amount of null varchars, those would naturally decrease the density of the vector.

**Comment by siddharthteotia** [comment]
addressed review comments, added unit tests.

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java [hunk 2]
```diff
@@ -169,6 +169,45 @@ public void setInitialCapacity(int valueCount) {
...
+   * Get the density of this ListVector
+   * @return density
+   */
+  public double getDensity() {
+    if (valueCount == 0) {
+      return 0.0D;
+    }
+    double totalDataSize = 0.0D;
+    for (int i = 0; i < valueCount; i++) {
+      if (isSet(i) == 1) {
```
Comment (jacques-n) [inline java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java]: this can be done in aggregate, no need for loop.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/BaseRepeatedValueVector.java [hunk 3]
```diff
@@ -143,6 +143,35 @@ public void setInitialCapacity(int numRecords) {
...
+   *                vector has a list of 10 values.
+   *                A density value of 0.1 implies out of 10 positions in
+   *                the list vector, 1 position has a list of size 1 and
+   *                remaining positions are null (no lists). This helps
+   *                in tightly controlling the memory we provision for
+   *                inner data vector.
+   */
+  public void setInitialCapacity(int numRecords, double density) {
+    offsetAllocationSizeInBytes = (numRecords + 1) * OFFSET_WIDTH;
+    final int innerValueCapacity = (int)(numRecords * density);
```
Comment (jacques-n) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/BaseRepeatedValueVector.java]: let's throw exception here if we're truncating.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java [hunk 4]
```diff
@@ -102,6 +97,60 @@ public void initializeChildrenFromFields(List<Field> children) {
...
+   * forces the memory requirement to go beyond what was needed.
+   *
+   * @param numRecords value count
+   * @param density density of ListVector. Density is the average size of
+   *                list per position in the List vector. For example, a
+   *                density value of 10 implies each position in the list
+   *                vector has a list of 10 values.
+   *                A density value of 0.1 implies out of 10 positions in
+   *                the list vector, 1 position has a list of size 1 and
+   *                remaining positions are null (no lists). This helps
```
Comment (jacques-n) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: null (no lists) or empty lists

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java [hunk 5]
```diff
@@ -102,6 +97,60 @@ public void initializeChildrenFromFields(List<Field> children) {
...
+   * Get the density of this ListVector
+   * @return density
+   */
+  public double getDensity() {
+    if (valueCount == 0) {
+      return 0.0D;
+    }
+    double totalListSize = 0.0D;
+    for (int i = 0; i < valueCount; i++) {
+      if (isSet(i) == 1) {
```
Comment (jacques-n) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: as above, no need for a loop.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/BaseRepeatedValueVector.java (hunk 3 above)
Comment (siddharthteotia) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/BaseRepeatedValueVector.java]: done.

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java (hunk 2 above)
Comment (siddharthteotia) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java]: done.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java (hunk 4 above)
Comment (siddharthteotia) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: done.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java (hunk 5 above)
Comment (siddharthteotia) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: done.

**Comment by siddharthteotia** [comment]
addressed review comments

File: java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java [hunk 6]
```diff
@@ -169,6 +169,42 @@ public void setInitialCapacity(int valueCount) {
...
+   */
+  public void setInitialCapacity(int valueCount, double density) {
+    final long size = (long) (valueCount * density);
+    if (size < 1) {
+      throw new IllegalArgumentException("With the provided density and value count, potential capacity of the data buffer is 0");
+    }
+    if (size > MAX_ALLOCATION_SIZE) {
+      throw new OversizedAllocationException("Requested amount of memory is more than max allowed");
+    }
+    valueAllocationSizeInBytes = (int) size;
```
Comment (vkorukanti) [inline java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java]: Do we need to round this up to nearest 2^n?

Comment (siddharthteotia) [reply java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java]: It happens during actual allocation. setInitialCapacity() doesn't allocate but just records the state for amount of memory we want to allocate.
